ambry-store/src/test/java/com.github.ambry.store/BlobStoreStatsTest.java

```diff
@@ -241,8 +315,351 @@ public void testValidDataSizeAfterDeletes() throws InterruptedException, StoreEx
+    CountDownLatch secondCheckpointLatch = new CountDownLatch(1);
+    CountDownLatch secondCheckpointHoldLatch = new CountDownLatch(1);
+    int throttleCountAtInterest = state.referenceIndex.size() + 1;
+    throttlers.put(BlobStoreStats.IO_SCHEDULER_JOB_TYPE,
+        new MockThrottler(scanStartedLatch, scanHoldLatch, secondCheckpointLatch, secondCheckpointHoldLatch,
+            throttleCountAtInterest));
+    BlobStoreStats blobStoreStats =
+        setupBlobStoreStats(2 * (int) (state.time.milliseconds() / BUCKET_SPAN_IN_MS), state.time.milliseconds());
+    // proceed only when the scan is started
+    assertTrue("IndexScanner took too long to start", scanStartedLatch.await(5, TimeUnit.SECONDS));
+    // 1 new put that will expire in 30 seconds
+    long expiresAtInMs = state.time.milliseconds() + TimeUnit.SECONDS.toMillis(30);
+    List<IndexEntry> newPutEntries = state.addPutEntries(1, CuratedLogIndexState.PUT_RECORD_SIZE, expiresAtInMs);
+    // 6 new puts with no expiry
+    newPutEntries.addAll(state.addPutEntries(6, CuratedLogIndexState.PUT_RECORD_SIZE, Utils.Infinite_Time));
+    for (IndexEntry entry : newPutEntries) {
+      blobStoreStats.handleNewPutEntry(entry.getValue());
+    }
+    List<MockId> newDeletes = new ArrayList<>();
+    // 1 delete from the first index segment
+    newDeletes.add(state.getIdToDeleteFromIndexSegment(state.referenceIndex.firstKey()));
+    // 1 delete from the last index segment
+    newDeletes.add(state.getIdToDeleteFromIndexSegment(state.referenceIndex.lastKey()));
+    for (MockId idToDelete : newDeletes) {
+      if (idToDelete != null) {
+        newDelete(blobStoreStats, idToDelete);
+      }
+    }
+    // continue the scan towards the first checkpoint
+    scanHoldLatch.countDown();
+    assertTrue("IndexScanner took too long to start the second checkpoint",
+        secondCheckpointLatch.await(5, TimeUnit.SECONDS));
+    // add more entries after the second checkpoint is taken
+    // 3 new puts with no expiry
+    newPutEntries = state.addPutEntries(3, CuratedLogIndexState.PUT_RECORD_SIZE, Utils.Infinite_Time);
+    for (IndexEntry entry : newPutEntries) {
+      blobStoreStats.handleNewPutEntry(entry.getValue());
+    }
+    newDeletes.clear();
+    // 1 delete from the last index segment
+    newDeletes.add(state.getIdToDeleteFromIndexSegment(state.referenceIndex.lastKey()));
+    for (MockId idToDelete : newDeletes) {
+      if (idToDelete != null) {
+        newDelete(blobStoreStats, idToDelete);
+      }
+    }
+    // continue the scan towards the second checkpoint
+    secondCheckpointHoldLatch.countDown();
+    // a probe put with a latch to inform us about the state of the queue
+    CountDownLatch queueProcessedLatch = new CountDownLatch(1);
+    blobStoreStats.handleNewPutEntry(new MockIndexValue(queueProcessedLatch, state.index.getCurrentEndOffset()));
+    assertTrue("QueueProcessor took too long to process the new entries",
```

Comment:
If you mean by accounting the 0 size probe blob then yes. However, blobs prior to the probe blob is guaranteed to be accounted for since QueueProcessor process entries sequentially from a queue.